### PR TITLE
prov/shm: fix locking when processing response from self

### PR DESCRIPTION
**Pull request by aingerson**

### CODE REVIEW

File: prov/shm/src/smr_progress.c [hunk 1]
```diff
@@ -138,8 +139,11 @@ static int smr_progress_resp_entry(struct smr_ep *ep, struct smr_resp *resp,
 			"unidentified operation type\n");
 	}
 
-	if (fastlock_tryacquire(&peer_smr->lock))
-		return -FI_EAGAIN;
+	if (strcmp(smr_name(peer_smr), smr_name(ep->region))) {
```
Comment (shefty) [inline prov/shm/src/smr_progress.c:142]: This is an expensive comparison to insert into the fast path.  I would find a different way to compare.  Maybe compare the shared memory region pointers?

Comment (aingerson) [reply prov/shm/src/smr_progress.c:142]: Would the pointers be the same if we mapped our own region again?

Comment (aingerson) [reply prov/shm/src/smr_progress.c]: Answer is apparently yes! Updated.

File: prov/shm/src/smr_progress.c [hunk 2]
```diff
@@ -66,6 +66,7 @@ static int smr_progress_resp_entry(struct smr_ep *ep, struct smr_resp *resp,
 	struct smr_inject_buf *tx_buf = NULL;
 	struct smr_sar_msg *sar_msg = NULL;
 	uint8_t *src;
+	bool locked = false;
```
Comment (shefty) [inline prov/shm/src/smr_progress.c:69]: You can drop this now and just repeat the pointer comparison.  I would add a short comment to the code regarding why the lock isn't required as well.

File: prov/shm/src/smr_progress.c (hunk 1 above)
Comment (aingerson) [reply prov/shm/src/smr_progress.c]: Ugh nevermind, I switched the if to only lock if they are the same 🙄 The comparison fails. Any other ideas?

**Comment by aingerson** [comment]
@shefty Updated with the new method we discussed. This seems to work. I'll do more testing to verify though.

**Comment by aingerson** [comment]
@shefty Updated again with ep_name_list locking

File: prov/util/src/util_shm.c [hunk 3]
```diff
@@ -208,11 +216,28 @@ int smr_map_create(const struct fi_provider *prov, int peer_count,
...
 	struct smr_region *peer;
 	size_t size;
 	int fd, ret = 0;
+	struct dlist_entry *entry;
+
+	pthread_mutex_lock(&ep_list_lock);
+	entry = dlist_find_first_match(&ep_name_list, smr_match_name,
+				       peer_buf->peer.name);
+	pthread_mutex_unlock(&ep_list_lock);
+	if (entry) {
```
Comment (shefty) [inline prov/util/src/util_shm.c:236]: As long as you're adding locks around the ep_name_list, I would continue to hold the lock while entry is being accessed.  (There's no reference being held on entry.)

Comment (aingerson) [reply prov/util/src/util_shm.c:235]: Updated

File: prov/util/src/util_shm.c [hunk 4]
```diff
@@ -208,11 +216,29 @@ int smr_map_create(const struct fi_provider *prov, int peer_count,
...
+	pthread_mutex_lock(&ep_list_lock);
+	entry = dlist_find_first_match(&ep_name_list, smr_match_name,
+				       peer_buf->peer.name);
+	if (entry) {
+		peer_buf->region = container_of(entry, struct smr_ep_name,
+						entry)->region;
+		pthread_mutex_unlock(&ep_list_lock);
+		return FI_SUCCESS;
+	}
+	pthread_mutex_unlock(&ep_list_lock);
```
Comment (shefty) [inline prov/util/src/util_shm.c:241]: There should be a similar check somewhere to handle the unmapping.

File: prov/util/src/util_shm.c [hunk 5]
```diff
@@ -321,11 +347,19 @@ int smr_map_add(const struct fi_provider *prov, struct smr_map *map,
...
 		return;
 
-	munmap(map->peers[id].region, map->peers[id].region->total_size);
+	pthread_mutex_lock(&ep_list_lock);
+	entry = dlist_find_first_match(&ep_name_list, smr_match_name,
+				       map->peers[id].peer.name);
+	if (!entry)
+		munmap(map->peers[id].region, map->peers[id].region->total_size);
+	pthread_mutex_unlock(&ep_list_lock);
+
```
Comment (aingerson) [inline prov/util/src/util_shm.c:362]: @shefty Like this?

Comment (shefty) [reply prov/util/src/util_shm.c:362]: I think?  You can drop the lock after the find call here, since you're not accessing entry.  : D

Comment (aingerson) [reply prov/util/src/util_shm.c]: So picky! Updated :)

**Pull request merged**